components/raftstore/src/store/fsm/peer.rs

```diff
@@ -987,36 +987,95 @@ where
+    // proposed to persist the state in the RegionLocalState to lock the region.
+    // Once this command is submitted, a callback will be invoked to call msg
+    // WaitApplyFlashback to complete the stop read stop write stop scheduling.
+    fn on_prepare_flashback(&mut self, cb: Callback<EK::Snapshot>) {
+        let region_id = self.region().get_id();
         info!(
             "prepare flashback";
-            "region_id" => self.region().get_id(),
+            "region_id" => region_id,
+            "peer_id" => self.fsm.peer.peer_id(),
+        );
+        // check persist state
+        let state_key = keys::region_state_key(region_id);
+        if let Some(target_state) = self
+            .ctx
+            .engines
+            .kv
+            .get_msg_cf::<RegionLocalState>(CF_RAFT, &state_key)
+            .unwrap()
+        {
+            info!(
+                "Prepare Flashback, already has a flashback local state:{:?} for {:?} ",
+                target_state.get_flashback_state(),
+                self.region(),
+            );
+            if target_state.get_flashback_state() == FlashbackState::Processing {
+                // check memory state
+                if self.fsm.peer.flashback_state.is_none() {
+                    info!("region is already has a flashback state in memory");
+                    let cmd_resp = RaftCmdResponse::default();
+                    cb.invoke_with_response(cmd_resp);
+                }
+                return;
+            }
+        }
+        // set persist state
+        let req = {
+            let mut request = new_admin_request(region_id, self.fsm.peer.peer.clone());
+            let mut admin = AdminRequest::default();
+            admin.set_cmd_type(AdminCmdType::PrepareFlashback);
+            request.set_admin_request(admin);
+            request
+        };
+        self.propose_raft_command(req, cb, DiskFullOpt::AllowedOnAlmostFull);
+    }
+
+    // Call msg WaitApplyFlashback to stop the scheduling and RW tasks, and write
+    // the persistent state in memory. Once called, the main thread will wait
+    // for the callback to complete. WWe place a flag in the request, which is
+    // checked when the pre_propose_raft_command is called. Stopping tasks is
+    // done by applying the flashback-only command in this way, But for RW local
+    // reads which need to be considered, we let the leader lease to None to
+    // ensure that local reads are not executed.
+    fn on_wait_apply_flashback(&mut self, cb: Callback<EK::Snapshot>) {
+        let region_id = self.region().get_id();
+        info!(
+            "wait apply flashback";
+            "region_id" => region_id,
             "peer_id" => self.fsm.peer.peer_id(),
         );
         if self.fsm.peer.flashback_state.is_some() {
-            ch.send(false).unwrap();
+            cb.invoke_with_response(new_error(Error::FlashbackInProgress(region_id)));
```

Comment:
I don't get why we need to invoke an error if it's in a flashback state. To Wait for the application after the flashback state has been set should be expected.